controller/src/main/java/io/pravega/controller/server/bucket/BucketManager.java

```diff
@@ -0,0 +1,198 @@
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+import java.util.stream.IntStream;
+
+/**
+ * This class is central to bucket management. This is instantiated once per service type. It is responsible for acquiring
+ * ownership of buckets for the given service and then starting one bucket service instance for each bucket it owns. 
+ * It manages the lifecycle of bucket service.
+ * The implementation of this class implements store specific monitoring for bucket ownership changes.
+ * This service does not have any dedicated threads. Upon receiving any new notification, it submits a "tryTakeOwnership" 
+ * work in the executor's queue. 
+ */
+@Slf4j
+public abstract class BucketManager extends AbstractService {
+    private final String processId;
+    @Getter(AccessLevel.PROTECTED)
+    private final BucketStore.ServiceType serviceType;
+    private final Function<Integer, BucketService> bucketServiceSupplier;
+    private final ConcurrentMap<Integer, BucketService> buckets;
+    private final ScheduledExecutorService executor;
+
+    BucketManager(final String processId, final BucketStore.ServiceType serviceType, final ScheduledExecutorService executor,
+                  final Function<Integer, BucketService> bucketServiceSupplier) {
+        this.processId = processId;
+        this.serviceType = serviceType;
+        this.executor = executor;
+        this.buckets = new ConcurrentHashMap<>();
+        this.bucketServiceSupplier = bucketServiceSupplier;
+    }
+
+    @Override
+    protected void doStart() {
+        Futures.allOf(IntStream.range(0, getBucketCount()).boxed().map(x -> initializeBucket(x)
+                .thenCompose(v -> tryTakeOwnership(x))).collect(Collectors.toList()))
+                .thenAccept(x -> startBucketOwnershipListener())
+                .whenComplete((r, e) -> {
+                    if (e != null) {
+                        notifyFailed(e);
+                    } else {
+                        notifyStarted();
+                    }
+                });
+    }
+
+    protected abstract int getBucketCount();
+
+    private CompletableFuture<Void> tryTakeOwnership(int bucket) {
+        return takeBucketOwnership(bucket, processId, executor)
+                         .thenCompose(isOwner -> {
+                    if (isOwner) {
+                        log.info("{}: Taken ownership for bucket {}", serviceType, bucket);
+
+                        // Once we have taken ownership of the bucket, we will register listeners on the bucket. 
+                        CompletableFuture<Void> bucketFuture = new CompletableFuture<>();
+                        
+                        buckets.compute(bucket, (x, y) -> {
+                            if (y == null) {
+                                BucketService bucketService = bucketServiceSupplier.apply(bucket);
+
+                                bucketService.addListener(new Listener() {
```

Comment:
> This doesn't need to be done inside of compute. In fact it doesn't seem like a good idea because the contract allows for the compute function to be called multiple times.

Even if remapping function inside `compute` is called multiple times, there is a synchronization block under which it is called. so its only one remapping execution which means first time we compute the service and subsequent times it returns the previous value. 

the reason we want to do it inside `compute` method is we want to a) create only one instance of bucketService (call the supplier once) and call start on it exactly once. 
doing it outside compute may mean `service.startAsync` can be called multiple times and we need to safeguard against that. 
and start has to be called after registering the listener or else the listener doesnt receive any notificatoin about service started.

Reply:
The code currently will create more that one bucket service, and register more than one listener, and it will call startAsync more than once. This is what I meant by "it doesn't seem like a good idea because the contract allows for the compute function to be called multiple times".